### ISSUE TITLE
Rationale and problems for additional hash in InReplyTo

### ISSUE THREAD

**Issue by mar-v-in** [ietf-wg-mimi/draft-ietf-mimi-content#29]
What is the rationale of having both the `messageId` (which is the hash over the encrypted message) and a hash over the unencrypted message in the reply to?

As it is currently states, the hash is algorithm agnostic (has `hashAlg`). Which hash algorithms are allowed and how do clients discover which algorithm to use?

Additionally, the hash algorithm agnosticism implies that clients need to keep a copy of the unencrypted message content to be able to compute the hash to be able to compare (or pre-compute all possible hash algorithms). Most clients naturally don't want to keep such a copy (as they will extract the relevant message content into their own data structure and then abandon the raw unencrypted message content bytes) and pre-computing multiple possible hashes for every message (even those that never receive a reply) seems wasteful on resources. Additionally, keeping a copy of the message content would potentially be in conflict with message expiry.

**Comment by rohanmahy** [comment]
@kkohbrok  proposed hashing the plaintext with a sender-provided salt (in the message), and including that in InReplyTo.

Note we already have a client franking secret we could use for this.

**Comment by mar-v-in** [comment]
Thanks @kkohbrok for the detailed explanation.

The issue I see with your sketched history sharing:
- How does Alice learn and verify the original sender of past messages forwarded by Bob?
- How does Alice learn and verify that the set of past messages forwarded by Bob is complete and/or does not include messages that have not been part of the original conversation history?
- How does Alice learn and verify timestamps of the past messages forwarded by Bob?

The plaintext message content (at least as of last time I checked) does not include the sender id. Bob thus doesn't need to find a 2nd preimage to the hash to craft new messages that look like they've been replied to, he can just use the original and manipulate authorship.

To sketch an attack scenario:
1. Bob, Charlie and David in the room:
2. David: "Can I get access to resource X?" with nonce Y and hash Z
3. Alice joins room
4. Bob forwards messages, but replaces David's message with a message that looks as if it was sent by Bob with same content and nonce Y as David's message and thus same hash Z
5. Charlie: "[reply to hash Z] Sure. Alice, could you go ahead and grant him access."

So beside the nonce, the hash needs to at least also cover the message sender's id or the reply-to needs to reference the sender id.

**Comment by mar-v-in** [comment]
Besides all of this, it doesn't explain why we need hash algorithm agility within InReplyTo. If we were to always use the same hash algorithm we also use for the outer MessageId, it would just mean that every message gets a second message id (let's call it MessageContentId) that is generally suitable for your message history sharing usecase, but means significantly reduced complexity on clients (which would just need to compute this second message id once when receiving the message and then have it available for indexing). If we define MessageContentId to be `hash(sender id || mimiContent || nonce)` it can effectively replace most usecases of MessageId. As the purpose of MessageId within mimi-content is already rather unclear (see #34), we might just want to get rid of it entirely within mimi-content (can still be useful for mimi-protocol). In that case the `messageId` , `hash` and `alg` fields in the InReplyTo could be removed completely and replaced with only a `messageContentId`...

PS: Whatever we decide to use in InReplyTo should also be used in the delivery reports.

**Comment by kkohbrok** [comment]
Since we don't have a history sharing mechanism in MIM (yet) my main concern was to prevent leaking the replied-to plaintext.

That being said your points/questions are all very valid if we indeed want a history sharing mechanism baked into MIMI. The problem is that MLS (as MIMI's crypto-engine) doesn't really provide any of these guarantees (or make them easy to get). My intuition is that all of this is possible, but would require some rather complex additional cryptographic machinery. So before we start adding this complexity, I'd like some buy-in from the WG that we actually want a history sharing mechanism on a MIMI-level. Or at least one that doesn't require Alice to just trust Bob to a large degree.

If we do have that buy-in, I'd be happy to help design such a mechanism. Sounds like a fun problem to solve :-)

Meanwhile I think adding the sender id to the hash is a great idea to at least get some slightly better guarantees without much additional complexity.

**Comment by mar-v-in** [comment]
If we don't want a trust-less history sharing, we can get rid of the hash entirely to prevent leaking message content and only use the MessageId, which is known and verifiable to all regular recipients of a message, but completely useless and effectively random to anyone that wasn't in the group at the time. And using the MessageId when sharig history with trusted peers (e.g. multiple devices of the same user) is fine as well, as the MessageId as it was before decrypting can just be appended to each shared plaintext message as long as the original receiving client had it stored (which it needs to anyway to support referencing it).

So I agree we should first figure out if trust-less history sharing is a desired festure, also noting that adding it likely implies that it will increase the size of every message (at least for the nonce, but likely even more) and complexity for every client, even if it is not used.

**Comment by kkohbrok** [comment]
I wouldn't call it trust-less but having the hash there gives you at least some degree of security with relatively little overhead. We could still do that if we decide that we don't want a more complex solution.

I agree that whatever mechanism we come up with will increase the message size, but we could turn it on or off on a per-group basis and leave out the hash (or whatever else) if history sharing is turned off.

**Comment by mar-v-in** [comment]
I don't agree. If a malicious history sharer can send an arbitrary message history without the recipient being able to verify any of it, the only thing the hash provides is that if InReplyTo is used, the recipient can learn that they never received a specific message or that it was falsified - and this is only if that message receives a reply. This doesn't make history sharing from users/devices that you don't trust 100% any more reasonable than without it.

**Comment by kkohbrok** [comment]
The mechanism is not to protect the recipient of the history. It's to protect (to a certain degree) the sender of the replyTo who has not decided to trust anyone at this point. When they include the hash, they can at least be certain that recipients can verify what content (and even what sender) they meant to reply to.

**Comment by mar-v-in** [comment]
If the sender replies to a specific MessageId (which is the hash of the encrypted content) they can already be certain that the recipient can verify all of this, for as long as the recipient was a member of the group when said message was sent. If there was no history sharing, that's all that's needed, because that's everyone who has the message. Everyone that joined later will only be able to tell "That's a reply to a message I don't know about".

So the only purpose of this is history sharing which we haven't decided to be a thing in MIMI at all. I agree that if we had history sharing, something like this is likely needed, although details probably will be a direct consequence of the history sharing itself and don't need to be defined specifically for InReplyTo either (e.g. if we had history sharing, each message was to likely get an additional ID to the MessageId that is likely is the hash of something that includes the unencrypted message content and some structure of message metadata like sender and hub timestamp). That's what I meant earlier with that we can just have another MessageId that is then also used for delivery reports and and probably everywhere else where a reference to a previous message is desired (instead of using the current MessageId). I believe such ID would also be more useful for vcon for example (which effectively is some kind of history sharing).

**Comment by rohanmahy** [comment]
Can we close this as satisfied by #36 ?

**Comment by tgeoghegan** [comment]
MIMI WG bug scrub:
- Konrad: this only matters if there is some history syncing mechanism. We should revisit this question if/when we decide to do history syncing.
- Rohan: some of this is resolved by using the optional franking mechanism
- Tim to write a new issue on mimi-arch to track the question of whether MIMI wants a history syncing mechanism and link to this issue from there.

**Comment by rohanmahy** [comment]
Read through the thread. There are some requirement to harvest here at the point that we want a history mechanism, but the mechanism currently in *this* draft seems to solve everybody's concerns. Closing.
